src/node_large_page.cc

```diff
@@ -0,0 +1,233 @@
+// Copyright (C) 2018 Intel Corporation
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom
+// the Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included
+// in all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL
+// THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES
+// OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+// ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE
+// OR OTHER DEALINGS IN THE SOFTWARE.
+//
+// SPDX-License-Identifier: MIT
+
+#include <errno.h>
+#include <fcntl.h>  // _O_RDWR
+#include <limits.h>  // PATH_MAX
+#include <locale.h>
+#include <signal.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdint.h>
+#include <string.h>
+#include <sys/types.h>
+#include <sys/mman.h>
+
+#include <string>
+#include <fstream>
+#include <iostream>
+
+// The functions in this file map the text segment of node into 2M pages.
+// The algorithm is quite simple
+// 1. Find the text region of node binary in memory
+// 2. Move the text region to large pages
+
+extern char __executable_start;
+extern char __etext;
+extern char __nodetext;
+
+namespace node {
+namespace largepages {
+#define ALIGN(x, a)           (((x) + (a) - 1) & ~((a) - 1))
+#define PAGE_ALIGN_UP(x, a)   ALIGN(x, a)
+#define PAGE_ALIGN_DOWN(x, a) ((x) & ~((a) - 1))
+
+struct TextRegion {
+  void * from;
+  void * to;
+  int    totalHugePages;
+  int64_t  offset;
+  bool   found_text_region;
+};
+
+    static void printSystemError(int error) {
+      fprintf(stderr, "Hugepages WARNING: %s\n", strerror(error));
+      return;
+    }
+
+    static struct TextRegion find_node_text_region() {
+      FILE *f;
+      int64_t  start, end, offset, inode;
+      char perm[5], dev[6], name[256];
+      int ret;
+      const size_t hugePageSize = 2L * 1024 * 1024;
+      struct TextRegion nregion;
+
+      nregion.found_text_region = false;
+
+      f = fopen("/proc/self/maps", "r");
```

Comment:
catpure fopen failure

Reply:
If /proc/self/maps is not there things are really messed up.